peps/pep-0728.rst

```diff
@@ -0,0 +1,647 @@
+there can be no other keys in ``MovieFinal`` other than the known ones.
+
+Type Consistency
+----------------
+
+In addition to the set ``S`` of keys of the explicitly defined items, a
+TypedDict type that has the item ``__extra__: T`` is considered to have an
+infinite set of items that all satisfy the following conditions:
+
+- If ``__extra__`` is read-only
+
+    - The key's value type is consistent with ``T``
+
+    - The key is not in ``S``.
+
+- If ``__extra__`` is not read-only
+
+    - The key is non-required
+
+    - The key's value type is consistent with ``T``
+
+    - ``T`` is consistent with the key's value type
+
+    - The key is not in ``S``.
+
+For type checking purposes, let ``__extra__`` be a non-required pseudo-item to
+be included whenever "for each ... item/key" is stated in
+:pep:`the existing type consistency rules from PEP 705 <705#type-consistency>`,
+and we modify it as follows:
+
+    A TypedDict type ``A`` is consistent with TypedDict ``B`` if ``A`` is
+    structurally compatible with ``B``. This is true if and only if all of the
+    following are satisfied:
+
+    * For each item in ``B``, ``A`` has the corresponding key, unless the item
+      in ``B`` is read-only, not required, and of top value type
+      (``ReadOnly[NotRequired[object]]``). [Edit: Otherwise, if the
+      corresponding key with the same name cannot be found in ``A``, "__extra__"
+      is considered the corresponding key.]
+
+    * For each item in ``B``, if ``A`` has the corresponding key [Edit: or
+      "__extra__"], the corresponding value type in ``A`` is consistent with the
+      value type in ``B``.
+
+    * For each non-read-only item in ``B``, its value type is consistent with
+      the corresponding value type in ``A``. [Edit: if the corresponding key
+      with the same name cannot be found in ``A``, "__extra__" is considered the
+      corresponding key.]
+
+    * For each required key in ``B``, the corresponding key is required in ``A``.
+      For each non-required key in ``B``, if the item is not read-only in ``B``,
+      the corresponding key is not required in ``A``.
+      [Edit: if the corresponding key with the same name cannot be found in
+      ``A``, "__extra__" is considered to be non-required as the corresponding
+      key.]
+
+The following examples illustrate these checks in action.
+
+``__extra__`` puts various restrictions on additional items for type
+consistency checks::
+
+    class Movie(TypedDict):
+        name: str
+        __extra__: int | None
+
+    class MovieDetails(TypedDict):
+        name: str
+        year: NotRequired[int]
+    
+    details: MovieDetails = {"name": "Kill Bill Vol. 1", "year": 2003}
+    movie: Movie = details  # Not OK. While 'int' is consistent with 'int | None',
+                            # 'int | None' is not consistent with 'int'
+
+The value type of "year" is not found in ``Movie``, and that it is not
```

Comment:
I don't follow this sentence. It might be enough to just have the examples and skip this paragraph of text. (Rename the second `MovieDetails` to something else.)